express-mutations/server.ts

```diff
@@ -0,0 +1,102 @@
+import express from 'express';
+import pg from 'pg';
+import { ClientError, errorMiddleware } from './lib/index.js';
+
+const db = new pg.Pool({
+  connectionString: 'postgres://dev:dev@localhost/pagila',
+  ssl: {
+    // Allow non-SSL traffic to localhost
+    rejectUnauthorized: false,
+  },
+});
+
+const app = express();
+app.use(express.json());
+// Endpoint for testing
+app.get('/api/actors/:actorId', async (req, res, next) => {
+  try {
+    const { actorId } = req.params;
+    if (!Number.isInteger(+actorId)) {
+      throw new ClientError(400, `Non-integer actorId: ${actorId}`);
+    }
+    const sql = `
+      select * from "actors"
+      where "actorId" = $1;
+    `;
+    const params = [actorId];
+    const result = await db.query(sql, params);
+    const actor = result.rows[0];
+    if (!actor) throw new ClientError(404, `actor ${actorId} not found`);
+    res.json(actor);
+  } catch (err) {
+    next(err);
+  }
+});
+
+app.post('/api/actors', async (req, res, next) => {
+  const { firstName, lastName } = req.body;
+  if (firstName === undefined) {
+    throw new ClientError(404, 'first name not found.');
+  }
+  if (lastName === undefined) {
+    throw new ClientError(404, 'last name not found.');
+  }
+  try {
+    const sql = `
+    insert into "actors" ("firstName","lastName")
+    values ($1,$2)
+    returning *
+    `;
+    const result = await db.query(sql, [firstName, lastName]);
+    const actors = result.rows[0];
+    res.status(201).json(actors);
+  } catch (err) {
+    next(err);
+  }
+});
+
+app.put('/api/actors/:actorId', async (req, res, next) => {
+  const { firstName, lastName } = req.body;
+  const { actorId } = req.params;
+  try {
+    const sql = `
+  update "actors"
+  set "firstName" = $1,
+  "lastName" = $2
+  where "actorId" = $3
+  returning *
+  `;
+    if (actorId === undefined)
+      throw new ClientError(404, 'actorId does not exisit.');
+    const result = await db.query(sql, [firstName, lastName, actorId]);
+    const actorsEdit = result.rows[0];
```

Comment:
You need to check that `actorsEdit` is defined here, and throw a 404 if not.

Your check on line 69-70 is checking that the `actorId` was provided in the request. It is not checking if the actorId provided actually exists in the database. That's why you need to check that `actorsEdit` is defined. If the `actorId` provided in the request is not found, then the SQL will not actually update a row, and `result.rows[0]` will be undefined. That's how you can tell if the `actorId` is referring to a row that actually exists in the table.

Note that on line 70, you should be responding with a 400, not a 404. If actorId is undefined it means the request did not contain an actorId, so it was a bad request, which is why it should be a 400.